File: ios/Classes/QiniuLivePlugin.m
#import "QiniuLivePlugin.h"
#import "AudioLive.h"

static NSObject<FlutterBinaryMessenger>* messager = nil;

@interface QiniuLivePlugin()

@property (nonatomic,strong) AudioLive *audioLive;

@end

@implementation QiniuLivePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    messager = [registrar messenger];
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"qiniu_live"
            binaryMessenger:[registrar messenger]];
  QiniuLivePlugin* instance = [[QiniuLivePlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS" stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  }
  else if([@"publishAudio" isEqualToString:call.method]){
      
      [self openMicriphone];
      
      NSString* token         = call.arguments[@"token"];
      NSString* roomName      = call.arguments[@"room"];
      NSDictionary* userData  = call.arguments[@"user_data"];
      NSString* appId         = call.arguments[@"app_id"];

      self.audioLive          = [AudioLive shareInstance];
      self.audioLive.userData = userData;
      self.audioLive.appId    = appId;
      self.audioLive.roomName = roomName;
      self.audioLive.token    = token;
      
      [self.audioLive publishAudio];
    
      [self listenUserEvent];
  }
  else if([@"leaveRoom" isEqualToString:call.method]){
      self.audioLive          = [AudioLive shareInstance];
      [self.audioLive.engine leaveRoom];
  }  else if([@"unPublish" isEqualToString:call.method]){
      self.audioLive = [AudioLive shareInstance];
      [self.audioLive unpublish];
  }
  else if([@"muteAudio" isEqualToString:call.method]){
      self.audioLive = [AudioLive shareInstance];
      [self.audioLive.engine muteAudio:!self.audioLive.isMute];
      self.audioLive.isMute = !self.audioLive.isMute;
  }
  else if([@"speakerOn" isEqualToString:call.method]){
      self.audioLive = [AudioLive shareInstance];
      self.audioLive.engine.muteSpeaker = !self.audioLive.engine.isMuteSpeaker;
  }
  else {
    result(FlutterMethodNotImplemented);
  }
    
}

- (void)openMicriphone {
    
    if ([[AVAudioSession sharedInstance] respondsToSelector:@selector(requestRecordPermission:)]) {
        [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
            if (granted) {
                NSLog(@"允许使用麦克风！");
            }
            else {
                NSLog(@"不允许使用麦克风！");
            }
        }];
    }
}

- (void)listenUserEvent {
    
    NSString *channelName = @"qiniu_live_users";
    FlutterEventChannel *evenChannal = [FlutterEventChannel eventChannelWithName:channelName binaryMessenger:messager];
    
    [evenChannal setStreamHandler:self];
    
    NSLog(@"channelName:%@",channelName);
}

- (FlutterError* _Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:(FlutterEventSink)events{
    if (events) {
        self.audioLive.flutterBlock = events;
    }
    return nil;
}

- (FlutterError* _Nullable)onCancelWithArguments:(id _Nullable)arguments{
    return nil;
}

@end

File: android/src/main/java/com/xiaofa/qiniu_live/QiniuLivePlugin.java
package com.xiaofa.qiniu_live;

import android.content.Context;

import com.xiaofa.qiniu_live.live.AudioLive;

import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** QiniuLivePlugin */
public class QiniuLivePlugin implements MethodCallHandler {

//  public interface FlutterLiveCallback {
//    List getLiveUsers();
//
//  }
//  static FlutterLiveCallback flutterLiveCallback;

  private AudioLive audioLive;
  static Context context;

  static BinaryMessenger messenger;
  private EventChannel.EventSink eventCallback;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "qiniu_live");
    context = registrar.activeContext();
    messenger = registrar.messenger();
    channel.setMethodCallHandler(new QiniuLivePlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method){
      case "getPlatformVersion":
        result.success("Android " + android.os.Build.VERSION.RELEASE);
        break;
      case "publishAudio":
        setEventToFlutter();
        audioLive = AudioLive.getSingleton();
        audioLive.context = context;
        String token = call.argument("token");
        String room = call.argument("room");
        Map user_data = call.argument("user_data");
        String app_id = call.argument("app_id");
        audioLive.roomName = room;
        audioLive.appId = app_id;
        audioLive.userData = user_data;
        audioLive.roomToken = token;

        audioLive.publishAudio();

        audioLive.eventCallback = eventCallback;

        break;
      case "leaveRoom":
        audioLive.leaveRoom();
        break;
      case "unPublish":
        audioLive.unPublish();
        break;
      case "muteAudio":
        audioLive.onToggleMic();
        break;
      case "speakerOn":
        audioLive.onToggleSpeaker();
        break;
      default:
        result.notImplemented();
    }
  }


  void setEventToFlutter (){

    new EventChannel(messenger, "qiniu_live_users").setStreamHandler(
            new EventChannel.StreamHandler() {
              @Override
              // 这个onListen是Flutter端开始监听这个channel时的回调，第二个参数 EventSink是用来传数据的载体。
              public void onListen(Object arguments, EventChannel.EventSink events) {
                if (events != null){
                  eventCallback = events;
                }
              }

              @Override
              public void onCancel(Object arguments) {
                // 对面不再接收

              }
            }
    );
  }
}
